cirq/sim/density_matrix_simulator.py

```diff
@@ -0,0 +1,425 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Simulator for density matrices that simulates noisy quantum circuits."""
+
+import collections
+
+from typing import cast, Dict, Iterator, List, Type, Union
+
+import numpy as np
+
+from cirq import circuits, linalg, ops, protocols, study, value
+from cirq.sim import density_matrix_utils, simulator
+
+class DensityMatrixSimulator(simulator.SimulatesSamples,
+                             simulator.SimulatesIntermediateState):
+    """A simulator for density matrices and noisy quantum circuits.
+
+    This simulator can be applied on circuits that are made up of operations
+    that have:
+        * a `_channel_` method
+        * a `_mixture_` method for a probabilistic combination of unitary gates.
+        * a `_unitary_` method
+        * a `_has_unitary_` and `_apply_unitary_` method.
+        * measurements
+    That is, the circuit must have elements that follow on of the protocols:
+        * `cirq.SupportsChannel`
+        * `cirq.SupportsMixture`
+        * `cirq.SupportsApplyUnitary`
+        * `cirq.SupportsUnitary`
+    or is a measurement.
+
+    This simulator supports three types of simulation.
+
+    Run simulations which mimic running on actual quantum hardware. These
+    simulations do not give access to the density matrix (like actual hardware).
+    There are two variations of run methods, one which takes in a single
+    (optional) way to resolve parameterized circuits, and a second which
+    takes in a list or sweep of parameter resolver:
+
+        run(circuit, param_resolver, repetitions)
+
+        run_sweep(circuit, params, repetitions)
+
+    These methods return `TrialResult`s which contain both the measurement
+    results, but also the parameters used for the parameterized
+    circuit operations. The initial state of a run is always the all 0s state
+    in the computational basis.
+
+    By contrast the simulate methods of the simulator give access to the density
+    matrix of the simulation at the end of the simulation of the circuit.
+    These methods take in two parameters that the run methods do not: a
+    qubit order and an initial state. The qubit order is necessary because an
+    ordering must be chosen for the kronecker product (see
+    `DensityMatrixTrialResult` for details of this ordering). The initial
```

Comment:
Does the density matrix include from simulate include measurement results, does it act like the measurement results are unknown, or does it require that all measurements be terminal?